### ISSUE TITLE
Attempt to request storage access in Firefox

### ISSUE THREAD

**Issue by jonkoops** [keycloak/keycloak#23872]
### Description

As part of fixing #22839 the choice was made to keep the implementation simple and not to call [`requestStorageAccess()`](https://developer.mozilla.org/en-US/docs/Web/API/Document/requestStorageAccess) due to  the added unpredictable behavior.

Under this issue the following must be done:

- Add the [`allow-storage-access-by-user-activation`](https://developer.mozilla.org/en-US/docs/Web/API/Storage_Access_API#security_measures)  option to the [`sandbox`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe#sandbox) attribute of the iframes embedded by Keycloak JS
- Implement the rest of the steps as outlined by ["Checking and requesting storage access"](https://developer.mozilla.org/en-US/docs/Web/API/Storage_Access_API/Using#checking_and_requesting_storage_access).
- Verify this improves compatibility as per [the test matrix](https://github.com/keycloak/keycloak/pull/23393#issuecomment-1727784845).

**Comment by vukanac** [comment]
On Safari 16.6, we are facing issues which were fixed in 22.0.3 and re-introduced in 22.0.4.
[](https://github.com/keycloak/keycloak/blob/22.0.4/services/src/main/resources/org/keycloak/protocol/oidc/endpoints/login-status-iframe.html#L33C9-L40C10)
Also, document.hasStorageAccess() returns undefined.
With this we are not able to show login form in iframe.
```
[Error] Unhandled Promise Rejection: undefined
    setError (keycloak.js:1278)
    (anonymous function) (keycloak.js:237)
[Error] Origin https://subdomain.domain.com is not allowed by Access-Control-Allow-Origin. Status code: 401
[Error] XMLHttpRequest cannot load https://login-subdomain.domain.com/auth/realms/XXXXX/account due to access control checks.
[Error] Failed to load resource: Origin https://subdomain.domain.com is not allowed by Access-Control-Allow-Origin. Status code: 401 (account, line 0)
```
Or we are not doing things correctly :(

**Comment by jonkoops** [comment]
@vukanac I don't believe this is the same issue. It looks like Safari is blocking the iframe entirely due to the way the `Access-Control-Allow-Origin` header is configured on your Keycloak server. Judging by the log, I don't think `document.hasStorageAccess()` is ever called, as the iframe that contains it is never loaded in the first place.

**Comment by jonkoops** [comment]
Moving this to v23, as we will no longer be doing patch releases for v22.

**Comment by originalnichtskoenner** [comment]
This might be irrelevant at this point but I had the same problem and got rid of it by setting checkLoginIframe to false.
```
keycloak.init({
    checkLoginIframe: false
})
```
It might at least be worth a try.

**Comment by deanmaster** [comment]
hi @jonkoops  this is for my understanding, this means Keycloak will call API to see if it has access to the data or this API will trigger the popup for the user to ask for permission for accessing the data ?
I just want to understand if this is changing the UX of the user in the application.

I know that in many website there is an asking for accept cookies so user can decide if they allow this or not.

Does this answer from the user include the permission for Keycloak access the storage ?

**Comment by jonkoops** [comment]
I don't think we will be implementing explicit user consent, instead this issue is about doing a 'best effort' call to the API as outlined [on MDN](https://developer.mozilla.org/en-US/docs/Web/API/Storage_Access_API/Using#checking_and_requesting_storage_access) (sans explicit user interaction).

**Comment by deanmaster** [comment]
thanks a lot I understand now

**Comment by ch-lepp** [comment]
It seems, this is no longer just a firefox issue.
Chrome is also preparing to phase out 3rd party cookies: [link](https://developers.google.com/privacy-sandbox/3pcd/prepare/prepare-for-phaseout?hl=en)

> If your site uses third-party cookies it's time to take action as we approach their deprecation. Chrome has restricted third-party cookies for 1% of users from January 4th, 2024 to facilitate testing, and plans to ramp up to 100% of users from Q3, subject to addressing any remaining competition concerns of the [UK's Competition and Markets Authority (CMA)](https://www.gov.uk/cma-cases/investigation-into-googles-privacy-sandbox-browser-changes).

**Comment by jonkoops** [comment]
Yes, not much we can do about it. Chrome implements a similar proccess, so Keycloak JS will just fall back to a redirect flow if it detects 3rd party cookies are not supported.

**Comment by deanmaster** [comment]
@jonkoops can we change to use Partioned Cookie instead of regular cookie for Keycloak SSO Cookie ?

**Comment by jonkoops** [comment]
Let's discuss it in the issue you reported when I get back from PTO. The answer is kinda complicated.

**Comment by nastasya-sorokina** [comment]
@jonkoops Hi, Has the problem been fixed? In version 26.1.0 in Firefox it is still not possible to get cookies if `onLoad: 'check-sso'` and `checkLoginIframe: true`:(
`checkLoginIframe: true` is required for our project

**Comment by jonkoops** [comment]
We're limited by what browsers allow, likely the cookies are just getting blocked, there is nothing more we can do about it.
